pkg/apis/kf/v1alpha1/app_types.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+import duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+import "k8s.io/kubernetes/pkg/apis/core"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// App is a 12-factor application deployed to Knative. It encompasses source
+// code, configuration, and the current state of the application.
+type App struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec AppSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status AppStatus `json:"status,omitempty"`
+}
+
+// AppSpec is the desired configuration for an App.
+type AppSpec struct {
+
+	// UpdateRequests is a unique identifier for an AppSpecTemplate.
+	// Updating sub-values will trigger a new value.
+	UpdateRequests int `json:"updaterequests"`
+
+	// Source contains the source configuration of the App.
+	// +optional
+	SourceSpec `json:",inline"`
+
+	// Template is a PodSpec with additional restrictions.
+	// The image name is ignored.
+	// The Spec contains configuration for the App's Pod.
+	// (Env, Vars, Quotas, etc)
+	// +optional
+	Template core.PodSpec `json:"spec,omitempty"`
```

Comment:
I think this field should be named `template` so we don't repeat spec twice in a row. I think Knative injects an intermediary field where it might make sense to stick the UpdateRequests like:

```
spec:
  template:
    updaterequests: 1
    spec:
      ...
```